File: Blackjack Project/Blackjack.py

import random

class Card:
    def __init__(self, rank, suit):
        self.rank = rank
        self.suit = suit

class Deck:
    def __init__(self):
        self.cards = []
        # Create a deck of cards by iterating through ranks and suits
        for suit in ["Hearts", "Diamonds", "Clubs", "Spades"]:
            for rank in ["2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"]:
                self.cards.append(Card(rank, suit))
        self.shuffle()  # Shuffle the deck when initialized

    def shuffle(self):
        random.shuffle(self.cards)  # Shuffle the deck

    def deal(self):
        return self.cards.pop()  # Deal a card from the top of the deck

def get_hand_value(hand):
    value = 0
    aces = 0
    # Calculate the total value of the hand, accounting for Aces
    for card in hand:
        if card.rank.isdigit():
            value += int(card.rank)
        elif card.rank in ["J", "Q", "K"]:
            value += 10
        elif card.rank == "A":
            value += 11
            aces += 1
    # Adjust for Aces if the total value exceeds 21
    while value > 21 and aces > 0:
        value -= 10
        aces -= 1
    return value

def print_hand(hand):
    # Print each card in the hand and its value
    for card in hand:
        print(f"{card.rank} of {card.suit}, ", end="")
    print(f"Value: {get_hand_value(hand)}")

def player_turn(deck, player_hand):
    while True:
        print("Your hand: ", end="")
        print_hand(player_hand)
        choice = input("Do you want to (h)it or (s)tand? ")
        if choice == 'h':
            player_hand.append(deck.deal())  # Deal a card to the player
            if get_hand_value(player_hand) > 21:
                print("You bust!")  # Inform the player if they bust
                return False
        elif choice == 's':
            return True  # Player stands

def ai_turn(deck, ai_hand):
    # AI hits until hand value is 17 or higher
    while get_hand_value(ai_hand) < 17:
        ai_hand.append(deck.deal())  # Deal a card to the AI
    print("AI's hand: ", end="")
    print_hand(ai_hand)

def determine_winner(player_hand, ai_hand):
    player_value = get_hand_value(player_hand)
    ai_value = get_hand_value(ai_hand)

    # Print the final hands of the player and AI
    print("Your final hand: ", end="")
    print_hand(player_hand)
    print("AI's final hand: ", end="")
    print_hand(ai_hand)

    # Determine the winner based on hand values
    if player_value > 21:
        print("You bust! AI wins.")
    elif ai_value > 21 or player_value > ai_value:
        print("You win!")
    elif player_value < ai_value:
        print("AI wins.")
    else:
        print("It's a tie!")

def main():
    deck = Deck()  # Initialize the deck
    player_hand = [deck.deal(), deck.deal()]  # Deal initial cards to the player
    ai_hand = [deck.deal(), deck.deal()]  # Deal initial cards to the AI

    if player_turn(deck, player_hand):  # Player's turn
        ai_turn(deck, ai_hand)  # AI's turn

    determine_winner(player_hand, ai_hand)  # Determine the winner

if __name__ == "__main__":
    main()  # Run the game

File: Blackjack Project/BlackJack.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <algorithm>

using namespace std;

enum Suit { Hearts, Diamonds, Clubs, Spades };
const vector<string> SuitNames = { "Hearts", "Diamonds", "Clubs", "Spades" };
const vector<string> RankNames = { "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A" };
const vector<int> RankValues = { 2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 10, 10, 11 };

struct Card {
    int rank;
    Suit suit;
};

class Deck {
private:
    vector<Card> cards;
public:
    Deck() {
        for (int suit = Hearts; suit <= Spades; ++suit) {
            for (int rank = 0; rank < RankValues.size(); ++rank) {
                cards.push_back(Card{ rank, static_cast<Suit>(suit) });
            }
        }
        shuffle();
    }

    void shuffle() {
        srand(static_cast<unsigned int>(time(0)));
        std::random_shuffle(cards.begin(), cards.end());
    }

    Card deal() {
        Card dealtCard = cards.back();
        cards.pop_back();
        return dealtCard;
    }
};

int getHandValue(const vector<Card>& hand) {
    int value = 0;
    int aces = 0;
    for (const Card& card : hand) {
        value += RankValues[card.rank];
        if (RankNames[card.rank] == "A") {
            aces++;
        }
    }
    while (value > 21 && aces > 0) {
        value -= 10;
        aces--;
    }
    return value;
}

void printHand(const vector<Card>& hand) {
    for (const Card& card : hand) {
        cout << RankNames[card.rank] << " of " << SuitNames[card.suit] << ", ";
    }
    cout << "Value: " << getHandValue(hand) << endl;
}

bool playerTurn(Deck& deck, vector<Card>& playerHand) {
    char choice;
    while (true) {
        cout << "Your hand: ";
        printHand(playerHand);
        cout << "Do you want to (h)it or (s)tand? ";
        cin >> choice;
        if (choice == 'h') {
            playerHand.push_back(deck.deal());
            if (getHandValue(playerHand) > 21) {
                cout << "You bust!" << endl;
                return false;
            }
        } else if (choice == 's') {
            return true;
        }
    }
}

void aiTurn(Deck& deck, vector<Card>& aiHand) {
    while (getHandValue(aiHand) < 17) {
        aiHand.push_back(deck.deal());
    }
    cout << "AI's hand: ";
    printHand(aiHand);
}

void determineWinner(const vector<Card>& playerHand, const vector<Card>& aiHand) {
    int playerValue = getHandValue(playerHand);
    int aiValue = getHandValue(aiHand);

    cout << "Your final hand: ";
    printHand(playerHand);
    cout << "AI's final hand: ";
    printHand(aiHand);

    if (playerValue > 21) {
        cout << "You bust! AI wins." << endl;
    } else if (aiValue > 21 || playerValue > aiValue) {
        cout << "You win!" << endl;
    } else if (playerValue < aiValue) {
        cout << "AI wins." << endl;
    } else {
        cout << "It's a tie!" << endl;
    }
}

int main() {
    Deck deck;
    vector<Card> playerHand;
    vector<Card> aiHand;

    playerHand.push_back(deck.deal());
    playerHand.push_back(deck.deal());
    aiHand.push_back(deck.deal());
    aiHand.push_back(deck.deal());

    if (playerTurn(deck, playerHand)) {
        aiTurn(deck, aiHand);
    }

    determineWinner(playerHand, aiHand);

    return 0;
}